File: apps/ios/Pebbles/Services/ConnectionsService.swift
import Foundation
import Observation
import Supabase
import os

/// Mutual connections (M49). Every write goes through a `security definer`
/// RPC — the three tables carry SELECT-only RLS, so there is no client write
/// path by design. All five RPCs return scalar `jsonb`, so responses decode
/// directly from the body (no `.single()`, which is for `SETOF` functions).
///
/// The service also parks the token from an invite universal link that landed
/// before a session existed: `RootView` flushes `pendingInviteToken` once
/// `supabase.session` appears.
@Observable
@MainActor
final class ConnectionsService {
    private let supabase: SupabaseService

    private let logger = Logger(subsystem: "app.pbbls.ios", category: "connections")

    /// Token captured from a universal link. Held until the accept sheet can
    /// be presented — i.e. until there is a session to accept with.
    var pendingInviteToken: String?

    init(supabase: SupabaseService) {
        self.supabase = supabase
    }

    /// `get_connections()` returns a jsonb array, newest first.
    func list() async throws -> [Connection] {
        try await supabase.client
            .rpc("get_connections")
            .execute()
            .value
    }

    /// Returns the caller's live invite, minting one only when none exists.
    /// `rotate` revokes the live invite and issues a fresh token — the whole
    /// revocation surface (a shared link stays valid until then).
    func createInvite(rotate: Bool = false) async throws -> ConnectionInvite {
        try await supabase.client
            .rpc("create_connection_invite", params: CreateInviteParams(pRotate: rotate))
            .execute()
            .value
    }

    /// Anon-callable preview: who is inviting, before accepting (or signing up).
    /// Never raises — an unusable token comes back as a status, not an error.
    func preview(token: String) async throws -> InvitePreview {
        try await supabase.client
            .rpc("preview_connection_invite", params: ["p_token": token])
            .execute()
            .value
    }

    /// Accepting is the mutual consent. A repeat accept succeeds with
    /// `alreadyConnected` — re-scanning a shared QR is the normal case, not an
    /// error. A block in either direction surfaces as `invite_expired`.
    func accept(token: String) async throws -> AcceptInviteResult {
        try await supabase.client
            .rpc("accept_connection_invite", params: ["p_token": token])
            .execute()
            .value
    }

    /// Severs the connection for both sides. `block` additionally records a
    /// one-way block that stops the peer re-entering through a live invite.
    func remove(connectionId: UUID, block: Bool = false) async throws {
        try await supabase.client
            .rpc(
                "remove_connection",
                params: RemoveConnectionParams(pConnectionId: connectionId.uuidString, pBlock: block)
            )
            .execute()
            .value
    }

    /// Best-effort refresh used by incidental callers; failures are logged and
    /// swallowed rather than surfaced.
    func refreshIgnoringFailure() async -> [Connection] {
        do {
            return try await list()
        } catch {
            logger.error("connections refresh failed: \(error.localizedDescription, privacy: .private)")
            return []
        }
    }
}

// MARK: - Wire params

/// The RPCs take heterogeneous params (a bool, a uuid string), so these use
/// `Encodable` structs rather than the `[String: String]` form the older
/// single-argument RPC call sites use.
private struct CreateInviteParams: Encodable {
    let pRotate: Bool

    enum CodingKeys: String, CodingKey {
        case pRotate = "p_rotate"
    }
}

private struct RemoveConnectionParams: Encodable {
    let pConnectionId: String
    let pBlock: Bool

    enum CodingKeys: String, CodingKey {
        case pConnectionId = "p_connection_id"
        case pBlock = "p_block"
    }
}

// MARK: - Errors

/// Postgres raises short slugs (`invite_expired`, …) and PostgREST delivers
/// them inside the message body, so the mapping is a substring match on the
/// lowercased text — the same contract the glyph market uses.
enum ConnectionsError: Error, LocalizedError, Equatable {
    case missingSession
    case inviteNotFound
    case inviteUnusable
    case ownInvite
    case unknown

    static func from(_ error: Error) -> ConnectionsError {
        let message = error.localizedDescription.lowercased()
        if message.contains("cannot_accept_own_invite") { return .ownInvite }
        // `invite_expired` also covers revoked tokens and a block in either
        // direction — deliberately indistinguishable, so a block is never
        // confirmed by the accept path.
        if message.contains("invite_expired") { return .inviteUnusable }
        if message.contains("invite_not_found") { return .inviteNotFound }
        if message.contains("not_authenticated") { return .missingSession }
        return .unknown
    }

    var errorDescription: String? {
        switch self {
        case .missingSession: return String(localized: "Please sign in again.")
        case .inviteNotFound, .inviteUnusable:
            return String(localized: "This invite link isn't valid anymore. Ask for a fresh one.")
        case .ownInvite: return String(localized: "That's your own invite link.")
        case .unknown: return String(localized: "Something went wrong. Please try again.")
        }
    }
}

File: apps/android/app/src/main/kotlin/app/pbbls/android/services/ConnectionsService.kt
package app.pbbls.android.services

import androidx.annotation.StringRes
import app.pbbls.android.R
import app.pbbls.android.features.glyph.models.GlyphStroke
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The connections seam (#849) — what [ConnectionsViewModel], [InviteViewModel]
 * and [AcceptInviteViewModel] are tested against.
 */
interface ConnectionsServicing {
    suspend fun list(): List<Connection>

    suspend fun createInvite(rotate: Boolean = false): ConnectionInvite

    suspend fun preview(token: String): InvitePreview

    suspend fun accept(token: String): AcceptInviteResult

    suspend fun remove(
        connectionId: String,
        block: Boolean = false,
    )
}

/**
 * Mutual connections (M49) — ports iOS `ConnectionsService`.
 *
 * Every write goes through a `security definer` RPC: the three tables carry
 * SELECT-only RLS, so there is no client write path by design. All five RPCs
 * return scalar `jsonb`, so responses use `decodeAs()` (never a single-row
 * accessor). Errors propagate to the caller, which owns loading/error view
 * state; [connectionsErrorMessage] maps the Postgres slugs to copy.
 */
@Singleton
class ConnectionsService
    @Inject
    constructor(
        private val supabase: SupabaseService,
    ) : ConnectionsServicing {
        /** `get_connections() returns jsonb` — the caller's connections, newest first. */
        override suspend fun list(): List<Connection> =
            supabase.client.postgrest
                .rpc("get_connections")
                .decodeAs()

        /**
         * Returns the caller's live invite, minting one only when none exists.
         * [rotate] revokes the live invite and issues a fresh token — the whole
         * revocation surface, so a link already shared stays alive until then.
         */
        override suspend fun createInvite(rotate: Boolean): ConnectionInvite =
            supabase.client.postgrest
                .rpc(
                    "create_connection_invite",
                    buildJsonObject { put("p_rotate", rotate) },
                ).decodeAs()

        /**
         * Anon-callable preview: who is inviting, before accepting. Never raises —
         * an unusable token comes back as a status, not an error.
         */
        override suspend fun preview(token: String): InvitePreview =
            supabase.client.postgrest
                .rpc(
                    "preview_connection_invite",
                    buildJsonObject { put("p_token", token) },
                ).decodeAs()

        /**
         * Accepting is the mutual consent. A repeat accept SUCCEEDS with
         * `alreadyConnected` — re-scanning a shared QR is the normal case, not an
         * error. A block in either direction surfaces as `invite_expired`.
         */
        override suspend fun accept(token: String): AcceptInviteResult =
            supabase.client.postgrest
                .rpc(
                    "accept_connection_invite",
                    buildJsonObject { put("p_token", token) },
                ).decodeAs()

        /**
         * Severs the connection for both sides. [block] additionally records a
         * one-way block that stops the peer re-entering through a live invite.
         */
        override suspend fun remove(
            connectionId: String,
            block: Boolean,
        ) {
            supabase.client.postgrest.rpc(
                "remove_connection",
                buildJsonObject {
                    put("p_connection_id", connectionId)
                    put("p_block", block)
                },
            )
        }
    }

// ---------------------------------------------------------------------------
// Wire models. UUIDs are Strings on Android; timestamps stay Strings because
// the RPC projections are rendered relatively, never parsed for arithmetic.
// ---------------------------------------------------------------------------

/** The peer's avatar geometry — the render payload only, no id or ownership. */
@Serializable
data class PeerGlyph(
    val strokes: List<GlyphStroke> = emptyList(),
    @SerialName("view_box") val viewBox: String,
)

/** A peer as projected by the connection RPCs — never a `profiles` row. */
@Serializable
data class ConnectionPeer(
    @SerialName("display_name") val displayName: String? = null,
    val glyph: PeerGlyph? = null,
)

/** One row of `get_connections()`. */
@Serializable
data class Connection(
    @SerialName("connection_id") val connectionId: String,
    @SerialName("connected_at") val connectedAt: String,
    val peer: ConnectionPeer,
)

/** `create_connection_invite()` — the live invite, re-displayed on every open. */
@Serializable
data class ConnectionInvite(
    val token: String,
    @SerialName("expires_at") val expiresAt: String,
) {
    /**
     * The one canonical invite URL. No custom scheme: this https link opens
     * the app through App Links when installed, and the web accept page when
     * not (design D11).
     */
    val url: String get() = inviteUrl(token)
}

/** `preview_connection_invite()` — never an error, always a status. */
@Serializable
data class InvitePreview(
    val status: String,
    /** Present only for `valid`: a withdrawn or expired token goes fully dark. */
    val inviter: ConnectionPeer? = null,
) {
    val isValid: Boolean get() = status == "valid"
}

/** `accept_connection_invite()`. `alreadyConnected` is a success state. */
@Serializable
data class AcceptInviteResult(
    @SerialName("connection_id") val connectionId: String,
    @SerialName("already_connected") val alreadyConnected: Boolean = false,
    val peer: ConnectionPeer,
)

/** Composes the canonical invite URL for [token]. Pure; JVM-tested. */
fun inviteUrl(token: String): String = "https://$INVITE_HOST/invite/$token"

/**
 * Extracts the token from an invite App Link, accepting exactly the canonical
 * URL shape. A loose parser would hand arbitrary strings to the accept RPC.
 * Pure; JVM-tested.
 */
fun parseInviteToken(url: String?): String? {
    val uri = runCatching { java.net.URI(url ?: return null) }.getOrNull() ?: return null
    if (uri.scheme != "https" || uri.host != INVITE_HOST) return null
    val segments = uri.path.orEmpty().split('/')
    val parts = segments.filter { it.isNotEmpty() }
    if (parts.size != 2 || parts[0] != "invite") return null
    return parts[1].ifEmpty { null }
}

const val INVITE_HOST = "www.pbbls.app"

/**
 * A connections failure as its inline string resource (pure; JVM-tested).
 *
 * Was a substring scan of the exception text in the iOS order; it now matches
 * the server's own answer exactly — see [DataError] for why the old form was
 * reading the request URL as well as the error. iOS still has the substring
 * version (`ConnectionsError.from`), so the two surfaces have diverged here on
 * purpose until iOS gets the same treatment.
 */
@StringRes
fun connectionsErrorMessage(error: DataError): Int =
    when (error) {
        DataError.Network -> R.string.error_offline
        DataError.Unauthorized -> R.string.connections_error_session
        is DataError.Conflict ->
            when (error.code) {
                "cannot_accept_own_invite" -> R.string.connections_error_own_invite
                // `invite_expired` also covers revoked tokens and a block in either
                // direction — deliberately indistinguishable, so a block is never
                // confirmed by the accept path.
                "invite_expired", "invite_not_found" -> R.string.connections_error_invite_unusable
                "not_authenticated" -> R.string.connections_error_session
                else -> R.string.connections_error_generic
            }

        DataError.NotFound -> R.string.connections_error_invite_unusable
        DataError.Quota, is DataError.Unknown -> R.string.connections_error_generic
    }
